src/java/org/apache/cassandra/repair/autorepair/AutoRepairUtils.java

```diff
@@ -447,6 +452,182 @@ public static AutoRepairHistory getHostWithLongestUnrepairTime(RepairType repair
+     * The criteria for this is to find the node with the oldest last repair finish time of which none of its replicasare currently under repair.
+     * @return The most eligible host to repair or null if no candidates before and including this nodes' current repair status.
+     */
+    @VisibleForTesting
+    public static AutoRepairHistory getMostEligibleHostToRepair(CurrentRepairStatus currentRepairStatus, UUID myId)
+    {
+        // 1. Sort repair histories from oldest completed to newest
+        Stream<AutoRepairHistory> finishedRepairHistories = currentRepairStatus.historiesWithoutOnGoingRepair
+                                                            .stream()
+                                                            .sorted((history0, history1) -> {
+                                                                if (history0.lastRepairFinishTime == history1.lastRepairFinishTime)
+                                                                    return 0;
+                                                                else
+                                                                    return history0.lastRepairFinishTime < history1.lastRepairFinishTime ? -1 : 1;
+                                                            });
+
+        // 2. Optimization: Truncate repair histories after myId so we don't evaluate anything more recent as if we
+        // aren't interested in anything that isn't this node.
+        final AtomicBoolean myHistoryFound = new AtomicBoolean(false);
+        finishedRepairHistories = finishedRepairHistories.takeWhile((history) -> {
+            if (!myHistoryFound.get())
+            {
+                if (history.hostId.equals(myId))
+                {
+                    myHistoryFound.set(true);
+                }
+                return true;
+            }
+            return false;
+        });
+
+        // If there are any hosts with ongoing repair, filter the repair histories to not include nodes whose replicas
+        // are ongoing repair.
+        Set<UUID> hostUUIDs = Sets.union(currentRepairStatus.hostIdsWithOnGoingForceRepair, currentRepairStatus.hostIdsWithOnGoingRepair);
+        if (!hostUUIDs.isEmpty())
+        {
+            // 3. Extract InetAddresses for each UUID as replicas are identified by their address.
+            Map<InetAddressAndPort, UUID> hostsBeingRepaired = getBroadcastAddressToHostIdMap(hostUUIDs);
+
+            // 4. Collect unique replication strategies and group them up with their keyspaces.
+            Map<AbstractReplicationStrategy, List<String>> replicationStrategies = getReplicationStrategies();
+
+            // 5. Filter out repair histories who have a replica being repaired, note that this is lazy, given the stream
+            //    is completed using findFirst, it should stop as soon as the matching criteria is met.
+            finishedRepairHistories = finishedRepairHistories.filter((history) -> !hasReplicaWithOngoingRepair(history,
```

Comment:
Should we keep this configurable? By default, we filter out histories that have a replica being repaired. 
But there can be a situation in which a node might starve for a long time, specifically if _parallel_repair_percentage_ is high, and for some use cases, folks might want to run regardless. For those cases, an admin can disable this flag, and it would not filter anything here?

Reply:
:+1: it's something i considered as well.  Would be easy enough to add a config to allow suppressing the behavior and just if check here.   Will add that.

Reply:
Added `allow_parallel_replica_repair` and `allow_parallel_replica_repair_across_schedules` to make this behavior configurable